engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java

```diff
@@ -0,0 +1,36 @@
+package io.camunda.zeebe.engine.processing.processinstance;
+
+import io.camunda.zeebe.engine.api.TypedRecord;
+import io.camunda.zeebe.engine.processing.streamprocessor.CommandProcessor;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.StateWriter;
+import io.camunda.zeebe.engine.processing.streamprocessor.writers.Writers;
+import io.camunda.zeebe.engine.state.KeyGenerator;
+import io.camunda.zeebe.protocol.impl.record.value.processinstance.ProcessInstanceModificationRecord;
+import io.camunda.zeebe.protocol.record.intent.ProcessInstanceModificationIntent;
+
+public final class ProcessInstanceModificationProcessor
+    implements CommandProcessor<ProcessInstanceModificationRecord> {
+
+  private final StateWriter stateWriter;
+  private final KeyGenerator keyGenerator;
+
+  public ProcessInstanceModificationProcessor(
+      final Writers writers, final KeyGenerator keyGenerator) {
+    stateWriter = writers.state();
+    this.keyGenerator = keyGenerator;
+  }
+
+  @Override
+  public boolean onCommand(
+      final TypedRecord<ProcessInstanceModificationRecord> command,
+      final CommandControl<ProcessInstanceModificationRecord> controller) {
+    final ProcessInstanceModificationRecord record = command.getValue();
+
+    stateWriter.appendFollowUpEvent(
+        keyGenerator.nextKey(), ProcessInstanceModificationIntent.MODIFIED, record);
+
+    controller.accept(ProcessInstanceModificationIntent.MODIFIED, record);
```

Comment:
❓ @reviewer I am not sure if this is correct, so please correct me if it's not. My goal is to write the `MODIFIED` event and after send a `MODIFIED` response back to the client, like mentioned in the issue:

- write a MODIFIED event when receiving a MODIFY command
- send a MODIFIED response to the client

Reply:
Yes, `controller.accept` takes care of this for you. It writes that event and (depending on the returned boolean of your `onCommand` method implementation) sends it also as a response.

The reason for this is that you've implemented the `ProcessInstanceModificationProcessor` as a `CommandProcessor`. When this is registered in `ProcessEventProcessors` it uses an overloaded method of `TypedRecordProcessors.onCommand`, that wraps your `CommandProcessor` implementation with the `CommandProcessorImpl`. This is a wrapper that tries to generify what a command processor does (i.e. write an event, and send a response). 

Alternatively, `ProcessInstanceModificationProcessor` could implement `TypedRecordProcessor` directly. In that case the registration happen over another of the overloaded methods, avoiding the wrapper. (Note the wrapper is an implementation of `TypedRecordProcessor` itself).

The thing about this wrapper is that it adds indirection in an attempt to clarify what a command processor should do. However, it does so in a rather unclear way, by using this command control and depending on the return value of the implemented method and such. It also restricts the number of records that a command processor can write (at least through the command control mechanism).

With previous changes, we've removed some of the usages of this wrapper, but this effort hasn't been completed. IMO we should avoid it for new processors.

I recommend that you make `ProcessInstanceModificationProcessor` implement `TypedRecordProcessor` directly.

Reply:
So my current solution would then write the event twice? I'll change it like you said, but that's still good to know 😄

Reply:
Oh wow, you're right. I missed that completely 😆 Goes to show how confusing this wrapper is